src/System.Linq/src/System/Linq/Union.cs

```diff
@@ -124,19 +170,101 @@ private Set<TSource> FillSet()
                 return set;
             }
 
-            public TSource[] ToArray()
+            internal override IEnumerable<TSource> GetSourceEnumerable(int index)
             {
-                return FillSet().ToArray();
+                Debug.Assert(index == 0 || index == 1);
+                return index == 0 ? _first : _second;
             }
 
-            public List<TSource> ToList()
+            internal override UnionIterator<TSource> Union(IEnumerable<TSource> next)
             {
-                return FillSet().ToList();
+                return new UnionIteratorN<TSource>(this, next, 2);
             }
+        }
 
-            public int GetCount(bool onlyIfCheap)
+        private sealed class UnionIteratorN<TSource> : UnionIterator<TSource>
+        {
+            private readonly UnionIterator<TSource> _previous;
+            private readonly IEnumerable<TSource> _next;
+            private readonly int _sourceIndex;
+
+            public UnionIteratorN(UnionIterator<TSource> previous, IEnumerable<TSource> next, int sourceIndex)
+                : base(previous.Comparer)
             {
-                return onlyIfCheap ? -1 : FillSet().Count;
+                Debug.Assert(previous != null);
+                Debug.Assert(next != null);
+                Debug.Assert(sourceIndex > 1);
```

Comment:
Yes. For a second I was thinking this was a place where it differed from concat, but concatN always has this index > 2 because it specialises doubles and triples. Since this specialises only doubles, index > 1.